firmware/include/stepper_generator.hpp

```diff
@@ -0,0 +1,99 @@
+#pragma once
+#include <array>
+#include <vector>
+#include <cinttypes>
+#include <cmath>
+#include <stm32g4xx_hal.h>
+#include "hardware.hpp"
+#include "config.hpp"
+
+
+namespace cube_hw
+{
+
+enum class motor_state{
+    RESET,      // not initialized
+    IDLE,       // standby
+    READY,      // dma ready
+    SKIP,       // dma ready with 0 steps
+    BUSY,       // running    
+    VELOCITY    // in velocity mode
+};
+
+struct MotorPins {
+    TIM_HandleTypeDef& htim;
+    const uint16_t channel;
+    GPIO_TypeDef* dir_gpio;
+    const uint16_t dir_pin;
+
+    MotorPins(TIM_HandleTypeDef& htim, const uint16_t channel, GPIO_TypeDef* dir_gpio, const uint16_t dir_pin);
+    MotorPins(const MotorPins&) = delete;
+    MotorPins& operator=(const MotorPins&) = delete;
+};
+
+/// @brief generator of PWM for stepper motor
+class StepperGenerator {
+    const MotorPins& pins;
+    int dma_burst_padding;
+    uint32_t dma_burst_length;
+    const unsigned _steps_per_mm;
+    motor_state _state = motor_state::RESET;
+    float ratio = 1.0f;
+    motor_config config;
+
+    // Dma
+    std::vector<uint16_t> acceleration_mem;
```

Comment:
I would strongly prefer static allocation for this. Or atleast pre-allocate in constructor and then never resize

Reply:
The vector is now resized only once